modules/images/image-loading-optimization/storage/data.php

```diff
@@ -0,0 +1,311 @@
+<?php
+/**
+ * Metrics storage data.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Gets the freshness age (TTL) for a given page metric.
+ *
+ * When a page metric expires it is eligible to be replaced by a newer one if its viewport lies within the same breakpoint.
+ *
+ * @return int Expiration TTL in seconds.
+ */
+function ilo_get_page_metric_freshness_ttl() {
+	/**
+	 * Filters the freshness age (TTL) for a given page metric.
+	 *
+	 * @param int $ttl Expiration TTL in seconds.
+	 */
+	return (int) apply_filters( 'ilo_page_metric_freshness_ttl', DAY_IN_SECONDS );
+}
+
+/**
+ * Determines whether the current response can be optimized.
+ *
+ * Only search results are not eligible by default for optimization. This is because there is no predictability in
+ * whether posts in the loop will have featured images assigned or not. If a theme template for search results doesn't
+ * even show featured images, then this isn't an issue.
+ *
+ * @return bool Whether response can be optimized.
+ */
+function ilo_can_optimize_response() {
+	$able = ! is_search();
+
+	/**
+	 * Filters whether the current response can be optimized.
+	 *
+	 * @param bool $able Whether response can be optimized.
+	 */
+	return (bool) apply_filters( 'ilo_can_optimize_response', $able );
+}
+
+/**
+ * Gets the normalized query vars for the current request.
+ *
+ * This is used as a cache key for stored page metrics.
+ *
+ * TODO: For non-singular requests, consider adding the post IDs from The Loop to ensure publishing a new post will invalidate the cache.
+ *
+ * @return array Normalized query vars.
+ */
+function ilo_get_normalized_query_vars() {
+	global $wp;
+
+	// Note that the order of this array is naturally normalized since it is
+	// assembled by iterating over public_query_vars.
+	$normalized_query_vars = $wp->query_vars;
```

Comment:
You mean beyond filtering `query_vars` to register additional public query vars? I should think that anything that is populating the `query_vars` is going to be essential for varying the cache for stored detection data. It will also automatically exclude arbitrary query vars added by analytics, such as `utm_*`. Granted, themes and plugins may also customize templates based on arbitrary query parameters being read directly from `$_GET`, so in this case the cache could be polluted. It's impossible to know which query parameters are going to impact the rendering of the page and which are just there for client-side purposes.

Reply:
Right, good point, I guess you need to consider them all.